src/dataunion/DataUnion.ts

```diff
@@ -564,30 +590,109 @@ export class DataUnion {
+            if (waitUntilTransportIsComplete) {
+                log(`Waiting for balance=${balanceBefore.toString()} change (poll every ${pollingIntervalMs}ms, timeout after ${retryTimeoutMs}ms)`)
+                await until(async () => !(await getBalanceFunc()).eq(balanceBefore), retryTimeoutMs, pollingIntervalMs).catch((e) => {
+                    const msg = `Timeout: Bridge did not transport withdraw message as expected. Fix: DataUnion.transportMessage("${messageHash}")`
+                    throw e.message.startsWith('Timeout') ? new Error(msg) : e
+                })
+                return null
+            }
+
+            // instead of waiting, hand out the messageHash so that someone else might do the transport using it
+            return messageHash
+        }
+
+        const ambTr = await this.transportMessage(messageHash, pollingIntervalMs, retryTimeoutMs)
+        if (waitUntilTransportIsComplete) {
+            log(`Waiting for balance ${balanceBefore.toString()} to change`)
+            await until(async () => !(await getBalanceFunc()).eq(balanceBefore), retryTimeoutMs, pollingIntervalMs)
+        }
+        return ambTr
+    }
+
+    // TODO: this doesn't belong here. Transporting a message is NOT dataunion-specific and needs nothing from DataUnion.ts.
+    //       It shouldn't be required to create a DataUnion object to call this.
+    //       This belongs to the StreamrClient, and if the code is too DU-specific then please shove it back to Contracts.ts.
+    //       Division to transportMessage and Contracts.transportSignaturesForMessage is spurious, they should be one long function probably.
+    /**
+     * @returns null if message was already transported, ELSE the mainnet AMB signature execution transaction receipt
+     */
+    async transportMessage(messageHash: AmbMessageHash, pollingIntervalMs: number = 1000, retryTimeoutMs: number = 300000) {
+        const helper = this.getContracts()
+        const sidechainAmb = await helper.getSidechainAmb()
+        const mainnetAmb = await helper.getMainnetAmb()
+
+        log(`Waiting until sidechain AMB has collected required signatures for hash=${messageHash}...`)
+        await until(async () => helper.requiredSignaturesHaveBeenCollected(messageHash), pollingIntervalMs, retryTimeoutMs)
+
+        const message = await sidechainAmb.message(messageHash)
+        if (message === '0x') {
+            throw new Error(`Message with hash=${messageHash} not found`)
         }
-        log(`Waiting for balance ${balanceBefore.toString()} to change`)
-        await until(async () => !(await getBalanceFunc()).eq(balanceBefore), retryTimeoutMs, pollingIntervalMs)
-        return tr
+        const messageId = '0x' + message.substr(2, 64)
+
+        log(`Checking mainnet AMB hasn't already processed messageId=${messageId}`)
+        const alreadySent = await mainnetAmb.messageCallStatus(messageId)
+        const failAddress = await mainnetAmb.failedMessageSender(messageId)
```

Comment:
Maybe could do a pass on all this to find things like this that can probably be done in parallel e.g.:

```js
const [alreadySent, failAddress] = await Promise.all([
  mainnetAmb.messageCallStatus(messageId),
  mainnetAmb.failedMessageSender(messageId),
])
```